### PR TITLE
 snapToCurrentLayer correctly handles Area type

### PR DESCRIPTION
**Pull request by 3nids**
also  revert adding tolerance to QgsSnappingUtils::snapToCurrentLayer (added by error from a bad rebase)

### CODE REVIEW

File: src/core/qgssnappingutils.cpp [hunk 1]
```diff
@@ -185,6 +185,13 @@ static void _updateBestMatch( QgsPointLocator::Match &bestMatch, const QgsPointX
   {
     _replaceIfBetter( bestMatch, loc->nearestEdge( pointMap, tolerance, filter ), tolerance );
   }
+  if ( bestMatch.type() != QgsPointLocator::Vertex && bestMatch.type() != QgsPointLocator::Edge && ( type & QgsPointLocator::Edge ) )
```
Comment (wonder-sk) [inline src/core/qgssnappingutils.cpp]: Should read "type & QgsPointLocator::Area" ?

File: src/core/qgspointlocator.h [hunk 2]
```diff
@@ -228,6 +228,14 @@ class CORE_EXPORT QgsPointLocator : public QObject
      */
     Match nearestEdge( const QgsPointXY &point, double tolerance, QgsPointLocator::MatchFilter *filter = nullptr );
 
+    /**
+     * Find nearest area to the specified point - up to distance specified by tolerance
+     * Optional filter may discard unwanted matches.
+     * This will first perform a pointInPolygon and return first result.
+     * If no match is found and tolerance is not 0, it will return nearestEdge.
```
Comment (wonder-sk) [inline src/core/qgspointlocator.h]: Missing note "added in 3.0"

File: src/core/qgspointlocator.cpp [hunk 3]
```diff
@@ -883,6 +882,36 @@ QgsPointLocator::Match QgsPointLocator::nearestEdge( const QgsPointXY &point, do
   return m;
 }
 
+QgsPointLocator::Match QgsPointLocator::nearestArea( const QgsPointXY &point, double tolerance, MatchFilter *filter )
+{
```
Comment (wonder-sk) [inline src/core/qgspointlocator.cpp]: Maybe worth checking whether the layer type is polygon and return invalid match immediately if it isn't?

**Review by wonder-sk — COMMENTED** [review]
Looks good in general, just please add tests for nearestArea + new behavior of snapToCurrentLayer

File: src/core/qgspointlocator.cpp (hunk 3 above)
Comment (3nids) [reply src/core/qgspointlocator.cpp]: pointInPoint does it internally, so I don't think we're losing anything here

File: src/core/qgspointlocator.h (hunk 2 above)
Comment (3nids) [reply src/core/qgspointlocator.h]: fixed

File: src/core/qgspointlocator.cpp [hunk 4]
```diff
@@ -205,7 +204,7 @@ class QgsPointLocator_VisitorArea : public IVisitor
       QgsFeatureId id = d.getIdentifier();
       QgsGeometry *g = mLocator->mGeoms.value( id );
       if ( g->intersects( mGeomPt ) )
-        mList << QgsPointLocator::Match( QgsPointLocator::Area, mLocator->mLayer, id, 0, QgsPointXY() );
+        mList << QgsPointLocator::Match( QgsPointLocator::Area, mLocator->mLayer, id, 0, mGeomPt.asPoint() );
```
Comment (3nids) [inline src/core/qgspointlocator.cpp]: @wonder-sk I think that it should return the input point as match point. Do you agree?

Comment (wonder-sk) [reply src/core/qgspointlocator.cpp]: Agreed!

**Review by wonder-sk — APPROVED** [review]
+1 to merge.

If we had a test also for snapToCurrentLayer / snapToMap for type=Area that would be absolutely stunning :smile:

**Comment by 3nids** [comment]
will do the test asap, thanks for the review

**Pull request merged**